x-pack/plugins/security_solution/server/search_strategy/timeline/factory/events/all/helpers.test.ts

```diff
@@ -371,4 +399,173 @@ describe('#formatTimelineData', () => {
+      // @ts-expect-error hit is minimal
+      const hit: EventHit = {
+        fields: {
+          'foo.bar': ['baz'],
+          'foo.barBaz': ['foo'],
+        },
+      };
+
+      expect(buildObjectForFieldPath('foo.barBaz', hit)).toEqual({
+        foo: { barBaz: ['foo'] },
+      });
+    });
+
+    it('builds an array of objects from a nested field', () => {
+      // @ts-expect-error hit is minimal
+      const hit: EventHit = {
+        fields: {
+          foo: [{ bar: ['baz'] }],
+        },
+      };
+      expect(buildObjectForFieldPath('foo.bar', hit)).toEqual({
+        foo: [{ bar: ['baz'] }],
+      });
+    });
+
+    it('builds intermediate objects for nested fields', () => {
+      // @ts-expect-error nestedHit is minimal
+      const nestedHit: EventHit = {
+        fields: {
+          'foo.bar': [
+            {
+              baz: ['host.name'],
+            },
+          ],
+        },
+      };
+      expect(buildObjectForFieldPath('foo.bar.baz', nestedHit)).toEqual({
+        foo: {
+          bar: [
+            {
+              baz: ['host.name'],
+            },
+          ],
+        },
+      });
+    });
+
+    it('builds intermediate objects at multiple levels', () => {
+      expect(buildObjectForFieldPath('threat.indicator.matched.atomic', eventHit)).toEqual({
+        threat: {
+          indicator: [
+            {
+              matched: {
+                atomic: ['matched_atomic'],
+              },
+            },
+            {
+              matched: {
+                atomic: ['matched_atomic_2'],
+              },
+            },
+          ],
+        },
+      });
+    });
+
+    it('preserves multiple values for a single leaf', () => {
+      expect(buildObjectForFieldPath('threat.indicator.matched.field', eventHit)).toEqual({
+        threat: {
+          indicator: [
+            {
+              matched: {
+                field: ['matched_field', 'other_matched_field'],
+              },
+            },
+            {
+              matched: {
+                field: ['matched_field_2'],
+              },
+            },
+          ],
+        },
+      });
+    });
+
+    describe('multiple levels of nested fields', () => {
+      let nestedHit: EventHit;
+
+      beforeEach(() => {
+        // @ts-expect-error nestedHit is minimal
+        nestedHit = {
+          fields: {
+            'nested_1.foo': [
+              {
+                'nested_2.bar': [
+                  { leaf: ['leaf_value'], leaf_2: ['leaf_2_value'] },
+                  { leaf_2: ['leaf_2_value_2', 'leaf_2_value_3'] },
+                ],
+              },
+              {
+                'nested_2.bar': [
+                  { leaf: ['leaf_value_2'], leaf_2: ['leaf_2_value_4'] },
+                  { leaf: ['leaf_value_3'], leaf_2: ['leaf_2_value_5'] },
+                ],
+              },
+            ],
+          },
+        };
+      });
```

Comment:
You do not need the `let` or `beforeEach`
```suggestion
      // @ts-expect-error nestedHit is minimal
      const nestedHit: EventHit = {
        fields: {
          'nested_1.foo': [
            {
              'nested_2.bar': [
                { leaf: ['leaf_value'], leaf_2: ['leaf_2_value'] },
                { leaf_2: ['leaf_2_value_2', 'leaf_2_value_3'] },
              ],
            },
            {
              'nested_2.bar': [
                { leaf: ['leaf_value_2'], leaf_2: ['leaf_2_value_4'] },
                { leaf: ['leaf_value_3'], leaf_2: ['leaf_2_value_5'] },
              ],
            },
          ],
        },
      };
```

Reply:
So this `let/beforeEach` is my paranoid "reset test data" pattern, meant to address the case where some function under test permutes its arguments. Consider the following scenario:

```ts
describe('permuting test data', () => {
  const a = ['foo'];

  it('works once', () => {
    expect(a).toEqual(['foo']);
    a.push('bar');
  });

  it('then breaks', () => {
    expect(a).toEqual(['foo']);
  });
});
```

Contrived example? Yes, we're not _usually_ writing code with side effects like this, and should strive not to. However, it inevitably happens, and occasionally the solution is to "fix" the tests such that they work against the permuted data. And then we have both order-dependent and incorrect tests 😦 .

Given the fact that this suite was using that shared `eventHit` variable all over the place, and I had no idea how the code worked before writing these new tests, I opted to go this route and ensure that these new tests were "clean" and not subject to previous permutations.

Reply:
gotchya